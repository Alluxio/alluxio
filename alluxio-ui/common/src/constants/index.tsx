export * from './types/INavigationData';
export * from './types/IFileBlockInfo';
export * from './types/IFileInfo';
