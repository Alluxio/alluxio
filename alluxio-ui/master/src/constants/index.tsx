export * from './data/navigation';
export * from './data/headerNavigation';
export * from './data/footerNavigation';
