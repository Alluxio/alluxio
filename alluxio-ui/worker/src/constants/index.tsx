export * from './data/footerNavigation';
export * from './data/headerNavigation';
