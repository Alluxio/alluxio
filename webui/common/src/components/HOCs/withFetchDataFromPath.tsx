/*
 * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
 * (the "License"). You may not use this work except in compliance with the License, which is
 * available at www.apache.org/licenses/LICENSE-2.0
 *
 * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied, as more fully set forth in the License.
 *
 * See the NOTICE file distributed with this work for information regarding copyright ownership.
 */

import React, {ReactNode} from 'react';

import {parseQuerystring} from '../../utilities';
import {IRequest} from "../../constants";
import {getDisplayName} from "../../utilities/misc/getDisplayName";

interface IFetchDataPathState {
    end?: string;
    limit?: string;
    offset?: string;
    path?: string;
}

interface IFetchDataPathProps {
    location: {
        search: string;
    };
    refresh: boolean;
    fetchRequest: (req: IRequest) => void;
}

interface IHandlers {
    createInputChangeHandler: (stateKey: string, stateValueCallback: (value: string) => string | undefined)  => ((event: React.ChangeEvent<HTMLInputElement>) => void);
    createButtonHandler: (stateKey: string, stateValueCallback: (value?: string) => string | undefined) => ((event: React.MouseEvent<HTMLButtonElement>) => void);
}

export type IFetchDataPathType = IFetchDataPathState & IFetchDataPathProps & IHandlers & {queryStringSuffix: string};

export function withFetchDataFromPath<TWrappedComponentProps extends IFetchDataPathProps>(WrappedComponent: React.ComponentType<TWrappedComponentProps>) {
    class fetchDataFromPathHoc extends React.Component<TWrappedComponentProps, IFetchDataPathState> {
        constructor(props: TWrappedComponentProps) {
            super(props);

            let {path, offset, limit, end} = parseQuerystring(this.props.location.search);
            path = decodeURIComponent(path || '');
            offset = offset || '0';
            this.state = {end, limit, offset, path};

            this.createInputChangeHandler = this.createInputChangeHandler.bind(this);
            this.createButtonHandler = this.createButtonHandler.bind(this);
        }

        public componentWillMount() {
            const {path, offset, limit, end} = this.state;
            this.fetchData({path, offset, limit, end});
        }

        public componentDidUpdate(prevProps: TWrappedComponentProps) {
            const {refresh, location: {search}} = this.props;
            const {refresh: prevRefresh, location: {search: prevSearch}} = prevProps;
            if (search !== prevSearch) {
                let {path, offset, limit, end} = parseQuerystring(search);
                path = decodeURIComponent(path || '');
                offset = offset || '0';
                this.setState({path, offset, limit, end});
                this.fetchData({path, offset, limit, end});
            } else if (refresh !== prevRefresh) {
                const {path, offset, limit, end} = this.state;
                this.fetchData({path, offset, limit, end});
            }
        }

        public render(): (undefined | ReactNode) {
            let queryStringSuffix = Object.entries(this.state)
                .filter((obj: any[]) => ['offset', 'limit', 'end'].includes(obj[0]) && obj[1] != undefined)
                .map((obj: any) => `${obj[0]}=${obj[1]}`).join('&');
            queryStringSuffix = queryStringSuffix ? '&' + queryStringSuffix : queryStringSuffix;

            return <WrappedComponent {...this.props}
                                     {...this.state}
                                     createInputChangeHandler={this.createInputChangeHandler}
                                     createButtonHandler={this.createButtonHandler}
                                     queryStringSuffix={queryStringSuffix} />;
        }

        private fetchData(request: IRequest) {
            this.props.fetchRequest(request);
        }

        private createInputChangeHandler(stateKey: string, stateValueCallback: (value: string) => string | undefined) {
            return (event: React.ChangeEvent<HTMLInputElement>) => {
                const value = event.target.value;
                this.setState({...this.state, [stateKey]: stateValueCallback(value)});
            };
        }

        private createButtonHandler(stateKey: string, stateValueCallback: (value?: string) => string | undefined) {
            return (event: React.MouseEvent<HTMLButtonElement>) => {
                this.setState({...this.state, [stateKey]: stateValueCallback()});
            };
        }
    };
    (fetchDataFromPathHoc as React.ComponentType<any>).displayName = `withFetchDataFromPath(${getDisplayName(WrappedComponent)})`;
    return fetchDataFromPathHoc;
}
