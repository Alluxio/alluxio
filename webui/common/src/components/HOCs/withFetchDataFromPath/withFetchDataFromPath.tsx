/*
 * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
 * (the "License"). You may not use this work except in compliance with the License, which is
 * available at www.apache.org/licenses/LICENSE-2.0
 *
 * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied, as more fully set forth in the License.
 *
 * See the NOTICE file distributed with this work for information regarding copyright ownership.
 */

import React from 'react';
import {IRequest} from "../../../constants";
import {getDisplayName, parseQueryString} from "../../../utilities";

interface IFetchDataFromPathState {
    request: IRequest;
}

export interface IFetchDataFromPathProps {
    fetchRequest: (req: IRequest) => void;
    location: {
        search: string;
    };
    refresh: boolean;
}

export function withFetchDataFromPath<T extends IFetchDataFromPathProps>(WrappedComponent: React.ComponentType<T>) {
    class fetchDataFromPathHoc extends React.Component<T, IFetchDataFromPathState> {
        constructor(props: T) {
            super(props);

            this.state = {request: this.getParsedQuery(this.props.location.search)};

            this.updateRequestParameter = this.updateRequestParameter.bind(this);
        }

        public componentWillMount() {
            this.fetchData(this.state.request);
        }

        public componentDidUpdate(prevProps: T) {
            const {refresh, location: {search}} = this.props;
            const {refresh: prevRefresh, location: {search: prevSearch}} = prevProps;
            if (search !== prevSearch) {
                const request = this.getParsedQuery(search);
                this.setState({request: request});
                this.fetchData(request);
            } else if (refresh !== prevRefresh) {
                this.fetchData(this.state.request);
            }
        }

        public render() {
            return <WrappedComponent {...this.props}
                                     {...this.state}
                                     upateRequestParameter={this.updateRequestParameter}
                                     queryStringSuffix={this.getQueryStringSuffix()} />;
        }

        private fetchData(request: IRequest) {
            this.props.fetchRequest(request);
        }

        private updateRequestParameter(stateKey: string, value: string | undefined) {
            const updatedReq: IRequest = {...this.state.request, [stateKey]: value};
            this.setState({request: updatedReq});
        }

        private getQueryStringSuffix(): string {
            let queryStringSuffix = Object.entries(this.state.request)
                .filter((obj: any[]) => ['offset', 'limit', 'end'].includes(obj[0]) && obj[1] != undefined)
                .map((obj: any) => `${obj[0]}=${obj[1]}`)
                .join('&');
            queryStringSuffix = queryStringSuffix ? '&' + queryStringSuffix : queryStringSuffix;
            return queryStringSuffix;
        }

        private getParsedQuery(query: string): IRequest {
            let {path, offset, limit, end} = parseQueryString(query);
            path = decodeURIComponent(path || '');
            offset = offset || '0';
            return {path, offset, limit, end};
        }
    }
    (fetchDataFromPathHoc as React.ComponentType<any>).displayName = `withFetchDataFromPath(${getDisplayName(WrappedComponent)})`;
    return fetchDataFromPathHoc;
}
