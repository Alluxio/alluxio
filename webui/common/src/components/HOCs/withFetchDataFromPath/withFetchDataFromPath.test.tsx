/*
 * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
 * (the "License"). You may not use this work except in compliance with the License, which is
 * available at www.apache.org/licenses/LICENSE-2.0
 *
 * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied, as more fully set forth in the License.
 *
 * See the NOTICE file distributed with this work for information regarding copyright ownership.
 */

import {configure, shallow, ShallowWrapper} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import sinon, {SinonSpy} from 'sinon';
import {IFetchDataFromPathProps, withFetchDataFromPath} from "./withFetchDataFromPath";
import {IRequest} from "../../../constants";

configure({adapter: new Adapter()});

const WrappedComponent = () => <div>Wrapped</div>;
const EnhancedComponent = withFetchDataFromPath(WrappedComponent);

describe('withFetchData HOC', () => {
    let props: IFetchDataFromPathProps;

    describe('Shallow component', () => {
        let shallowWrapper: ShallowWrapper;

        beforeEach(() => {
            props = {
                location: {search: ''},
                refresh: false,
                fetchRequest: sinon.spy(() => {})
            };
            shallowWrapper = shallow(<EnhancedComponent {...props}/>);
        });

        it('Renders without crashing', () => {
            expect(shallowWrapper.length).toEqual(1);
        });

        it('Calls fetchRequest in componentWillMount', () => {
            sinon.assert.calledOnce(props.fetchRequest as SinonSpy);
        });

        it('Check encoded search string get parsed into state', () => {
            shallowWrapper.setProps({location: {search: '?path=%2Fexample%2Fnotes.txt&offset=123'}});
            sinon.assert.calledTwice(props.fetchRequest as SinonSpy);
            const req: IRequest = shallowWrapper.state('request');
            expect(req.path).toEqual('/example/notes.txt');
            expect(req.offset).toEqual('123');
        });

        it('Check decoded search string get parsed into state', () => {
            shallowWrapper.setProps({location: {search: '?path=/example/notes.txt&offset=123'}});
            sinon.assert.calledTwice(props.fetchRequest as SinonSpy);
            const req: IRequest = shallowWrapper.state('request');
            expect(req.path).toEqual('/example/notes.txt');
            expect(req.offset).toEqual('123');
        });

        it('Did not call fetchRequest in componentDidUpdate refresh=false', () => {
            shallowWrapper.setProps({refresh: false});
            sinon.assert.calledOnce(props.fetchRequest as SinonSpy);
        });

        it('Calls fetchRequest in componentDidUpdate refresh=true', () => {
            sinon.assert.calledOnce(props.fetchRequest as SinonSpy); // componentWillMount
            shallowWrapper.setProps({refresh: true});
            sinon.assert.calledTwice(props.fetchRequest as SinonSpy); // componentDidMount
        });

        it('Matches snapshot - renders WrappedComponent', () => {
            expect(shallowWrapper).toMatchSnapshot();
        });
    });
});
