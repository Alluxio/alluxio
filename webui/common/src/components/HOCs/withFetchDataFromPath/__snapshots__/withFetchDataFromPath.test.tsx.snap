// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withFetchData HOC Shallow component Matches snapshot - renders WrappedComponent 1`] = `
<WrappedComponent
  fetchRequest={[Function]}
  location={
    Object {
      "search": "",
    }
  }
  queryStringSuffix="&offset=0"
  refresh={false}
  request={
    Object {
      "end": undefined,
      "limit": undefined,
      "offset": "0",
      "path": "",
    }
  }
  upateRequestParameter={[Function]}
/>
`;
